libkbfs/coalescing_context.go

```diff
@@ -0,0 +1,110 @@
+package libkbfs
+
+import (
+	"golang.org/x/net/context"
+	"reflect"
+)
+
+// CoalescingContext allows many contexts to be treated as one.  It waits on
+// all its contexts' Context.Done() channels, and when all of them have
+// returned, this CoalescingContext is canceled. At any point, a context can be
+// added to the list, and will subsequently also be part of the wait condition.
+// TODO: add timeout channel in case there is a goroutine leak
+type CoalescingContext struct {
+	context.Context
+	closeCh  chan struct{}
+	doneCh   chan struct{}
+	mutateCh chan context.Context
+	selects  []reflect.SelectCase
+}
+
+func (ctx *CoalescingContext) loop() {
+	for {
+		chosen, val, _ := reflect.Select(ctx.selects)
+		switch chosen {
+		case 0:
+			// request to mutate the select list
+			newCase := val.Interface().(context.Context)
+			if newCase != nil {
+				ctx.addContextLocked(newCase)
+			}
+		case 1:
+			// Done
+			close(ctx.doneCh)
+			return
+		default:
+			// The chosen channel has been closed. Remove it from our select list.
+			ctx.selects = append(ctx.selects[:chosen], ctx.selects[chosen+1:]...)
+			// If we have no more selects available, the request is done.
+			if len(ctx.selects) == 2 {
+				close(ctx.doneCh)
+				return
+			}
+		}
+	}
+}
+
+func (ctx *CoalescingContext) addContextLocked(other context.Context) {
+	ctx.selects = append(ctx.selects, reflect.SelectCase{
+		Dir:  reflect.SelectRecv,
+		Chan: reflect.ValueOf(other.Done()),
+	})
+}
+
+// NewCoalescingContext creates a new CoalescingContext. The context _must_ be
+// canceled to avoid a goroutine leak.
+func NewCoalescingContext(parent context.Context) (*CoalescingContext, context.CancelFunc) {
+	ctx := &CoalescingContext{
+		Context:  context.Background(),
```

Comment:
We can do this, but we'll only be able to pass the tags to the server at the time the request is started. WDYT?